src/lib/communication/DY.Lib.Communication.RequestResponse.Lemmas.fst

```diff
@@ -234,49 +306,80 @@ val mk_comm_layer_response_nonce_proof:
   Lemma
   (requires
     trace_invariant tr /\
-    has_communication_layer_reqres_state_predicates a /\
+    has_communication_layer_reqres_state_predicate a /\
     has_communication_layer_reqres_crypto_predicates a /\
-    bytes_well_formed tr req_meta_data.key
+    bytes_well_formed tr req_meta_data.key // Can be derived from CommServerReceiveRequest event
   )
   (ensures (
     match mk_comm_layer_response_nonce req_meta_data usg tr with
     | (None, tr_out) -> trace_invariant tr_out
     | (Some nonce, tr_out) -> (
-      is_knowable_by (get_response_label tr_out req_meta_data) tr_out nonce
+      trace_invariant tr_out /\
+      is_knowable_by (get_response_label tr_out req_meta_data) tr_out nonce /\
+      get_label tr_out nonce `can_flow tr_out` get_label tr_out req_meta_data.key
     )
   ))
 let mk_comm_layer_response_nonce_proof #invs #a tr req_meta_data usg =
-  reveal_opaque (`%mk_comm_layer_response_nonce) (mk_comm_layer_response_nonce #a)
+  reveal_opaque (`%mk_comm_layer_response_nonce) (mk_comm_layer_response_nonce #a);
+  reveal_opaque (`%get_response_label) (get_response_label);
+  ()
+
+val mk_comm_layer_response_nonce_labeled_proof:
+  {|protocol_invariants|} ->
+  #a:Type -> {|comm_layer_reqres_config a|} ->
+  tr:trace ->
+  req_meta_data:comm_meta_data a -> usg:usage -> prin:label ->
+  Lemma
+  (requires
+    trace_invariant tr /\
+    has_communication_layer_reqres_state_predicate a /\
+    has_communication_layer_reqres_crypto_predicates a /\
+    bytes_well_formed tr req_meta_data.key // Can be derived from CommServerReceiveRequest event
+  )
+  (ensures (
+    match mk_comm_layer_response_nonce_labeled req_meta_data usg prin tr with
+    | (None, tr_out) -> trace_invariant tr_out
+    | (Some nonce, tr_out) -> (
+      trace_invariant tr_out /\
+      is_knowable_by (get_response_label tr_out req_meta_data) tr_out nonce /\
+      get_label tr_out nonce `can_flow tr_out` get_label tr_out req_meta_data.key
+    )
```

Comment:
Should the postcondition here also say something about the nonce being knowable by `prin`?